### ISSUE TITLE
Display > Window > Dpi: Allow Hidpi

### ISSUE THREAD

**Issue by okapii** [godotengine/godot#42085]
**Godot version:**
3.2.3 RC 6 Official, MacOS

**OS/device including version:**
- MacBook Pro (13", 2020)
- 2,3 GHz Quad-Core Intel Core i7
- 16 GB 3733 MHz LPDDR4X
- Intel Iris Plus Graphics 1536 MB

**Issue description:**
Checking "Allow Hidpi" in the project settings (Display > Window > Dpi) scales down the window size to half (?) its intended size.
It happens with newly created vanilla 3D-projects, too. So this seems not to be about some other conflicting setting.
This appears to be a regression as the current stable build does not have this problem. It has been around for a number of release candidates already.

**Steps to reproduce:**
Check "Allow Hidpi" in project settings (Display > Window > Dpi).

**Minimal reproduction project:**
Happy to provide the fresh test-project I just created but since this is a one click operation for a newly created project, I refrained from uploading anything at this point for now.

**Comment by Calinou** [comment]
Can you reproduce this in older 3.2.3 RCs and betas? You can download them [here](https://downloads.tuxfamily.org/godotengine/3.2.3/).

**Comment by okapii** [comment]
Just checked RC1. The issue was in that one already.

**Comment by Calinou** [comment]
@okapii Thanks! If you can build Godot from source, can you |bisect the regression](https://docs.godotengine.org/en/latest/community/contributing/bisecting_regressions.html)? The "good" commit is `3.2.2-stable` and the bad commit is `a24e30abd7b1bc226dc1231ef2b8eb5a9ee50df6`.

**Comment by okapii** [comment]
Sorry, not really equipped with infrastructure / knowledge on how to do that … :(
Could try finding the time to figure this out next weekend.

**Comment by okapii** [comment]
I just updated to the new Godot 3.2.3 and the issue exists in that one, too.

**Comment by pk-nb** [comment]
Hi there! I ran into this issue as well after updating a basic project to 3.2.3 and bisected the two commits above. I believe I've found the specific change:

```
commit e9ab41b71d59f5555d1171c75b2aa5a72b6e801b
Author: bruvzg <7645683+bruvzg@users.noreply.github.com>
Date:   Tue Jul 7 21:37:05 2020 +0300

    [macOS, 3.2] Implement seamless display scaling.

 core/bind/core_bind.cpp    |  10 ++++
 core/bind/core_bind.h      |   2 +
 core/os/os.h               |   2 +
 doc/classes/OS.xml         |  20 +++++++
 editor/editor_node.cpp     |   4 ++
 editor/editor_run.cpp      |  29 ++++++----
 editor/project_manager.cpp |   7 ++-
 platform/osx/os_osx.h      |  12 +---
 platform/osx/os_osx.mm     | 138 ++++++++++++++++++++++++++-------------------
 9 files changed, 143 insertions(+), 81 deletions(-)
```


While I am new to the codebase, I would suspect the code change in [`editor/editor_run.cpp`](editor/editor_run.cpp) adjustments for window size and scale does not properly factor in the "Test Width" and "Test Height" values set in the project settings.

**Comment by Calinou** [comment]
cc @bruvzg

**Comment by bruvzg** [comment]
What's "wrong window scale"? When hiDPI is enabled real screen resolution is used and the window with the same pixel size will appear smaller, it's done for handling multiple displays with different DPI.

Before #40201 scaled resolution was used for both hiDPI and loDPI and multi monitor setups were completely unusable.

**Comment by pk-nb** [comment]
Hey @bruvzg, thanks for the response. The sizing behavior seems to be much improved! Hoping to check my understanding now the behavior has changed a bit. Are the following statements correct?

* When "Allow HiDPI = false", size units (Width, Height, Test Width, Test Height) are [effectively in the unit of _points_](https://developer.apple.com/library/archive/documentation/GraphicsAnimation/Conceptual/HighResolutionOSX/Explained/Explained.html).
* When "Allow HiDPI = true", size units (Width, Height, Test Width, Test Height) are effectively in the unit of _pixels_ of the highest DPI screen.

The only thing I'm worried about this change is that I'd like to use `Allow HiDPI = true" (for better font oversampling/rasterization) yet have a predictable window size. That seems impossible to do as we can't assume the device density (could be 1x, 2x, 3x, etc) but Test Width and Test Height are interpreted in pixels (effectively baking in a single density assumption). 

If my understanding is correct, could I suggest the following?

* "Test Width" and "Test Height" are always interpreted in the unit of _points_ and scaled by the screen density.
* (Width and Height are still always in pixels)
* When Test Width and Test Height are 0, the engine defaults to using the native size as _points_. 
     * That way the default settings yield 1 game pixel = 1 screen point = 1/2/3 actual pixels depending on the display.
     * Example: a game setup with `width=480, height=270, testWidth=0, testHeight=0` would yield a screen size of 480x270 points (rendering as 960x540pixels on a typical retina screen).

My feeling is that this would be more intuitive as it more closely matches how macOS handles window units. It seems it would make it easier for developers to setup the same window sizing in hiDPI mode too, without resorting to scripting.

I am new to this and there may be other solutions or thinking I'm unaware of—so forgive me if this suggestion seems like a bad idea. Thanks so much for your thoughts!

---

(aside, wanted to share the current behavior for reference)

### Example of behavior in 3.2.3 for reference

Screenshots of a game with a native size of 480x270 on a HiDPI screen.

<details>
<summary>Starting window size when "Allow HiDPI = false"</summary>

Window is 480x270 points (960x540 pixels on my HiDPI display).

<img width="592" alt="Screen Shot 2020-09-20 at 2 51 49 PM" src="https://user-images.githubusercontent.com/1911028/93719569-6c880e00-fb51-11ea-880b-c403a7a9cfba.png">

</details>


<details>
<summary>Starting window size when "Allow HiDPI = true"</summary>

Window is 240x135 points (480x270 pixels on my HiDPI display).

<img width="352" alt="Screen Shot 2020-09-20 at 2 51 36 PM" src="https://user-images.githubusercontent.com/1911028/93719593-93464480-fb51-11ea-9d16-14ef94e51aed.png">

</details>

**Comment by bruvzg** [comment]
Godot do not have any dynamic UI scaling support (#21446) and all window management is designed to use pixels, so it's always using pixels, we can't use points unless DPI independent rendering is implemented in the renderer.
In loDPI mode it's still using pixels but reports to OS that it does not support hiDPI.

Current solution was done to make editor usable on dual hiDPI + loDPI monitor setup (which is probably the most common mac setup, retina MacBook + external display). I agree, this approach is not most intuitive for macOS users, but it's consistent with other OSes.
And 3.2.2 behavior was my previous attempt to solve the issue (which was faulty, because I was implementing it without hiDPI monitor to test).

Using points is the best solution, but it will require much bigger changes to the renderer to make everything function properly (and outside of mine expertise).
And it's also the way to go on Windows 10, some Linux DEs (AFAIK only latest version of GNOME) seems to have DPI independent mode, but I'm sure how exactly it's implemented.

**Comment by Calinou** [comment]
> some Linux DEs (AFAIK only latest version of GNOME) seems to have DPI independent mode, but I'm sure how exactly it's implemented.

As far as I know, loDPI fallbacks on Linux are only available in some Wayland-based desktop environments, not X11. This is because X11 is too limited to provide proper mouse input for windows scaled by the display server.

**Comment by DoctorWhoof** [comment]
For what it's worth, I like the current behavior in 3.2.3, i.e. turning on HiDPI causes the window to be smaller, since it's displaying the true size **in pixels**.

This allows me the design a UI with hiDPI that looks identical, albeit smaller, to its non-hiDPI counterpart until I maximize it or go full screen.

What the OP is asking actually makes sense and is more in line with Apple's Retina screen guidelines (I think), but I find less practical. Maybe both can be an option?

**Comment by Calinou** [comment]
@DoctorWhoof We would really prefer to encourage best practices out of the box. Supporting hiDPI displays correctly is a best practice :slightly_smiling_face:

**Comment by DoctorWhoof** [comment]
> This already exists in the form of the 2d stretch mode

That works well for games, not for applications, which is what I'm focusing on now. For instance, when using 2D stretch mode all text is resized when resizing the window, which is not what you want in an application.

I feel that to satisfy the original request with Godot applications, Godot's Controls would need to behave in a resolution independent way - i.e. not use pixels as the unit. The effect can actually be achieved via code (I have it working, including resizing the UI on the fly with "zoom" controls), but it's a bit of a pain, especially if you didn't start the project like that from the get go.

Thanks!

**Comment by Calinou** [comment]
@DoctorWhoof As I said, we need to expose a way to manually change the scale factor independently of the window size. This is what https://github.com/godotengine/godot/pull/21446 provides.

**Comment by DoctorWhoof** [comment]
Yes @Calinou! Sorry I missed that point, I believe proper hiDPI (like the original request) + a global scale factor like that would work.

Thanks!

**Comment by okapii** [comment]
Has there been made any progress regarding this issue? I am pretty much stuck on 3.2.2 for utilising 2D-nodes in my 3D game for the GUI which are all displayed substantially too small in newer versions because of how things are handled now.
This seems like a fairly common combination so I would assume that I am probably not the only one having trouble here?

**Comment by Calinou** [comment]
> Has there been made any progress regarding this issue? 

No, my pull request needs to be remade for `4.0` and most likely won't make it to 3.2.x.

**Comment by okapii** [comment]
@Calinou Thanks for the feedback! Gonna give that a try!

**Comment by akien-mga** [comment]
As I understand it things are working as intended, but it's maybe not the most self explanatory behavior? Does it behave the same on macOS and on other OSes? Should we just document this better?

**Comment by pseidemann** [comment]
hi everyone :)

@akien-mga,
I think it's not intended once you are trying to create a gui application (like the godot editor itself).

imho there are multiple aspects which need to be considered.

**what is allow hidpi currently doing?**
it disables scaling of the viewport by the os. it also enables rendering it at the native resolution. this implies that when e.g. a system has 2x scaling (example: display is 4k but user wants desktop size to be logical 1080p), the project will be rendered 2x too small when not doing anything special about that.

**what are gui applications?**
a gui application is usually a project which uses mostly only control nodes. examples would be the godot editor itself or your favorite browser.

**how should gui applications behave regarding scaling?**
gui applications should have a proper control element size when starting the app and this size must stay the same regardless of the window size (different from games where fullscreen is often desired so the window size can be "predicted").

**what is the behavior with default settings?**
gui applications are properly scaled, though it could be argued that they don't render at native resolution, which MIGHT look blurry on some systems.
also I was told that there is some windows bug which is not fixable by godot which breaks window scaling (especially on multi monitor setups?).

**gui application: allow hidpi + fix scaling?**
I was told to enable stretch mode 2d but this is however NOT a solution because this breaks a fundamental principle how gui applications work. this would make control sizes dependent of the window size. see **how should gui applications behave regarding scaling?**.

I also ran into that issue while doing an unrelated fix for a godot demo but I couldn't get a good answer to that problem yet: https://github.com/godotengine/godot-demo-projects/pull/791

my conclusion is that currently "allow hidpi" is broken for gui applications. in fact that is also what the guide "multiple resolutions" is stating: https://docs.godotengine.org/en/stable/tutorials/rendering/multiple_resolutions.html

imho a fix might be worth here. maybe implementing dynamic scaling like the godot editor does.

another source: this person on q&a also explains the issue very well but the given answer doesn't fix the issues I mentioned here:
https://godotengine.org/qa/89180/whats-the-right-way-to-support-hidpi

**Comment by pseidemann** [comment]
still a problem in godot 4
